airflow-core/src/airflow/api_fastapi/core_api/services/public/hitl_shared_links.py

```
@@ -0,0 +1,382 @@
+        params_input: Parameters input for respond links
+        expires_at: Custom expiration time in hours
+    :param session: Database session (required to get task instance UUID)
+
+    :return: jwt-encoded token
+    """
+    token_data: HITLSharedLinkData = HITLSharedLinkData(
+        ti_id=ti_id,
+        dag_id=dag_id,
+        dag_run_id=dag_run_id,
+        task_id=task_id,
+        map_index=map_index,
+        shared_link_config=shared_link_config,
+    )
+
+    encoded_json = json.dumps(token_data).encode()
+    aesgcm = AESGCM(secret_key.encode())
+    iv = os.urandom(12)
+    ciphertext = aesgcm.encrypt(iv, encoded_json, None)
+    return base64.urlsafe_b64encode(iv + ciphertext).decode("utf-8")
+
+
+def decode_shared_data(
+    *,
+    secret_key: str,
+    token: str,
+) -> HITLSharedLinkData:
+    """
+    Decode a shared link token.
+
+    :param token: Base64-encoded token
+    :return: Decoded token data
+    :raises ValueError: If token is invalid or expired
+    """
+    raw = base64.urlsafe_b64decode(token.encode())
+    if len(raw) < 13:
+        raise ValueError("Token too short")
+
+    iv, ciphertext = raw[:12], raw[12:]
+    aesgcm = AESGCM(secret_key.encode())
+    plaintext = aesgcm.decrypt(iv, ciphertext, None)
+    return json.loads(plaintext)
+
+
+def _infer_base_url(base_url: str) -> str:
+    if not base_url:
+        base_url = conf.get("api", "base_url")
+        if not base_url:
+            raise ValueError("API base_url is not configured")
+    return base_url
+
+
+def _validate_redirect_config(
+    *, chosen_options: list[str] | None, params_input: dict[str, Any] | None
+) -> None:
+    if chosen_options or params_input:
+        raise ValueError(
+            '"chosen_options" and "params_input" should be empty "when "link_type" is set to "redirect"'
+        )
+
+
+def _validate_respond_config(*, chosen_options: list[str] | None) -> None:
+    if not chosen_options:
```

Should we handle the case where HITL isn’t set to `multiple`, but `len(chosen_options)` is greater than 1?

---

Sounds like a good idea 👀 My original thought was to defer it to task execution so that we can reduce one db query. but fail early seems to be better

---

done